source/common/common/inline_map.h

```diff
@@ -0,0 +1,515 @@
+   * Heterogeneous lookup is supported here.
+   * @param key the key to erase.
+   * @return the number of elements erased.
+   */
+  template <class GetKey> uint64_t erase(const GetKey& key) {
+    if (absl::optional<uint64_t> entry_id = inlineLookup(key); entry_id.has_value()) {
+      return clearInlineMapEntry(*entry_id);
+    } else {
+      return dynamic_entries_.erase(key);
+    }
+  }
+
+  /**
+   * Erase the entry by the given handle. If the handle is not valid, do nothing.
+   * @param handle the handle to erase.
+   * @return the number of elements erased.
+   */
+  uint64_t erase(Handle handle) { return clearInlineMapEntry(handle.inlineId()); }
+
+  /**
+   * Iterate all elements in the map.
+   * @param callback the callback function to handle each element.
+   */
+  void iterate(std::function<bool(const Key&, const Value&)> callback) const {
+    for (const auto& entry : dynamic_entries_) {
+      if (!callback(entry.first, entry.second)) {
+        return;
+      }
+    }
+
+    const uint64_t inline_keys_num = descriptor_.inlineKeysNum();
+    const auto& inline_keys = descriptor_.inlineKeys();
+    for (uint64_t i = 0; i < inline_keys_num; ++i) {
+      if (!inlineEntryValid(i)) {
+        continue;
+      }
+
+      if (!callback(inline_keys[i], inline_entries_[i])) {
+        return;
+      }
+    }
+  }
+
+  /**
+   * Clear all elements in the map.
+   */
+  void clear() {
+    // Destruct all inline entries.
+    const uint64_t inline_keys_num = descriptor_.inlineKeysNum();
+    for (uint64_t i = 0; i < inline_keys_num; ++i) {
+      clearInlineMapEntry(i);
+    }
+
+    // Clear the normal map.
+    dynamic_entries_.clear();
+  }
+
+  /**
+   * @return the number of elements in the map.
+   */
+  uint64_t size() const { return dynamic_entries_.size() + inline_entries_size_; }
+
+  /**
+   * @return true if the map is empty.
+   */
+  bool empty() const { return size() == 0; }
+
+  ~InlineMap() { clear(); }
+
+  // Override operator [] to support the normal key assignment.
+  template <class GetKey> Value& operator[](const GetKey& key) {
+    if (absl::optional<uint64_t> entry_id = inlineLookup(key); entry_id.has_value()) {
+      // This key is registered as inline key and try to add the value to the inline array.
+      if (!inlineEntryValid(*entry_id)) {
+        resetInlineMapEntry(*entry_id, Value());
+      }
+      return inline_entries_[*entry_id];
+    } else {
+      // This key is not registered as inline key and try to add the value to the normal map.
+      return dynamic_entries_[key];
+    }
+  }
+
+  // Override operator [] to support the handle of inline key assignment.
+  Value& operator[](Handle handle) {
+    if (!inlineEntryValid(handle.inlineId())) {
+      resetInlineMapEntry(handle.inlineId(), Value());
+    }
+    return inline_entries_[handle.inlineId()];
+  }
+
+  static std::unique_ptr<InlineMap> create(TypedInlineMapDescriptor& registry) {
```

Comment:
s/registry/descriptor/

add doxygen doc. Call out that this will implicitly finalize the descriptor if necessary, which (IIUC) is why it needs to be non-const.

Reply:
nice catch. thanks.